### DISCUSSION TITLE
Testing exception in plugin

### DISCUSSION THREAD

**Discussion by onosendi** [fastify/fastify#4573] [Q&A] [answered]
What's the proper way to test if an exception is thrown in a plugin?

```js
async function myPlugin(fastify, config) {
  if (config === 'foo') {
    throw new Error('foo');
  }

  fastify.decorate('foo', 'foo');
}

test('Throws foo', () => {
  const fastify = Fastify();

  // This doesn't work.
  expect(() => {
    fastify.register(myPlugin, 'foo');
  }).toThrow();
});
```

**Comment by Uzlopak** [comment, marked as answer]
With tap it would be

`await t.rejects(() => fastify.register(myPlugin, 'foo'), new Error('foo'))`

You are using mocha/chai. So read the documentation of your test runner on how to test for promises and rejections. 
e.g. https://www.chaijs.com/plugins/chai-as-promised/

**Reply by onosendi** [reply]
Yeah, I don't think jest and fastify are playing well together in this regard.

**Reply by Uzlopak** [reply]
You should not use jest for backend anyway.

But here is the link on how to test async code with jest

https://jestjs.io/docs/tutorial-async

**Reply by onosendi** [reply]
Yeah, the `try {} catch {}` seemed to have worked. Thank you.

**Reply by Uzlopak** [reply]
Just a recommendation. If you use try {} catch {} then either ensure the amount of assertions or throw an error in the the try block

If you dont do this, then it could be that your code does not throw anymore, but your tests pass.

**Reply by onosendi** [reply]
I'm doing:

```js
test('', async () => {
  try {
    await fastify.register(myPlugin, 'foo');
  } catch (e) {
    expect(e.message).toBe('foo');
  }
});
```

**Reply by onosendi** [reply]
Oh, hmm, I see what you're saying.

**Reply by Uzlopak** [reply]
```js
test('', async () => {
  try {
    await fastify.register(myPlugin, 'foo');
    fail('should have thrown');
  } catch (e) {
    expect(e.message).toBe('foo');
  }
});
```

Thats why you should use rejects.
https://jestjs.io/docs/expect#rejects

**Reply by onosendi** [reply]
Ah, yes. `rejects`/`resolves` is a lot cleaner. Thank you.

**Reply by onosendi** [reply]
If anyone else comes across this:

```js
test('', async () => {
  await expect(fastify.register(myPlugin, 'foo'))
    .rejects
    .toThrow('foo');
});
```
